### PR TITLE
Feature: Allow set specific template

### PR DESCRIPTION
**Pull request by luizdesign**
Proposal to resolve issue #170

### CODE REVIEW

**Comment by luizdesign** [comment]
Working to fix broken unit test

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,2039 @@
+{
```
Comment (vigneshshanmugam) [inline package-lock.json]: can you remove the package-lock file? we use `yarn.lock`

File: lib/fetch-template.js [hunk 2]
```diff
@@ -26,44 +26,78 @@ class TemplateError extends Error {
...
-        const pathname = url.parse(request.url, true).pathname;
-        const templatePath = path.join(templatesPath, pathname) + '.html';
-
-        return readFile(templatePath)
-            .then((baseTemplate) => {
-                if (typeof baseTemplateFn !== 'function') {
-                    return parseTemplate(baseTemplate);
-                }
+        return getTemplatePath(request, templatesPath)
+            .then((templatePath) => {
```
Comment (vigneshshanmugam) [inline lib/fetch-template.js]: can you combine it by this way
```js
getTemplatePath(request, templatesPath)
    .then(readFile)
    .then(baseTemplate)
```
looks cleaner IMO.

**Comment by vigneshshanmugam** [comment]
Thanks a lot for working on this. If you are interested, feel free to use https://github.com/ljharb/util.promisify to promisify the functions. We thought of using that package anyways.

File: lib/fetch-template.js [hunk 3]
```diff
@@ -26,44 +26,78 @@ class TemplateError extends Error {
...
 /**
+ * Returns the template path validating a exactly file or a directory
+ *
+ * @param  {Object} request - Request Object
+ * @param  {String} templatesPath - TemplatesPath config
+ *
+ * @return {Promise} Template Path on success or TemplateError on fail
+ */
+const getTemplatePath = (request, templatesPath) => new Promise((resolve, reject) => {
+    fs.lstat(
```
Comment (mroderick) [inline lib/fetch-template.js]: If you're using node 8.x, you can use [`util.promisify`](https://nodejs.org/api/util.html#util_util_promisify_original) to make this code simpler

**Comment by mroderick** [comment]
> If you are interested, feel free to use https://github.com/ljharb/util.promisify to promisify the functions. We thought of using that package anyways.

@vigneshshanmugam is [`util.promisify` ](https://nodejs.org/api/util.html#util_util_promisify_original) not available?

**Comment by vigneshshanmugam** [comment]
We still support node 6, so we can use that polyfill for node < 8

File: lib/fetch-template.js (hunk 2 above)
Comment (luizdesign) [inline lib/fetch-template.js]: Done! (:

File: package-lock.json (hunk 1 above)
Comment (luizdesign) [inline package-lock.json]: Done! (:

File: lib/fetch-template.js (hunk 3 above)
Comment (luizdesign) [inline lib/fetch-template.js]: I followed the code style of 'readFile' function, however we can change the both functions to promisify sintaxe. What do you think? Or do we open a new issue to discuss about it?

Comment (vigneshshanmugam) [inline lib/fetch-template.js]: We can also do it in a separate PR.

Comment (luizdesign) [inline lib/fetch-template.js]: Ok @vigneshshanmugam, I will open a issue about it.

Comment (luizdesign) [inline lib/fetch-template.js]: https://github.com/zalando/tailor/issues/172

File: lib/fetch-template.js [hunk 4]
```diff
@@ -26,35 +26,66 @@ class TemplateError extends Error {
...
  * @param {function=} baseTemplateFn - Function that returns the Base template name for a given page
  */
-module.exports = (templatesPath, baseTemplateFn) => 
+module.exports = (templatesPath, baseTemplateFn) =>
     (request, parseTemplate) => {
-        const pathname = url.parse(request.url, true).pathname;
-        const templatePath = path.join(templatesPath, pathname) + '.html';
-
-        return readFile(templatePath)
+        return getTemplatePath(request, templatesPath)
```
Comment (vigneshshanmugam) [inline lib/fetch-template.js]: Two problems with this approach

1. the url parsing happens twice with this approach. In `getTemplatePath` when its a directory and inside `baseTemplateFn`. 
2. You would also need to look at modify this logic https://github.com/zalando/tailor/blob/83446134c35dc5482bc48aa24463e2123d350285/lib/fetch-template.js#L63 Will totally break if the templatesPath is a single file instead of dir.

Comment (luizdesign) [inline lib/fetch-template.js]: Interesting, I will check this points as soon as possible!

Comment (luizdesign) [inline lib/fetch-template.js]: @vigneshshanmugam I adjust the duplicated url parsing and refactoring the logic to factory the complete template path. Now I working for break the process in case of single file

Comment (luizdesign) [inline lib/fetch-template.js]: @vigneshshanmugam Done!

File: lib/fetch-template.js [hunk 5]
```diff
@@ -26,45 +26,100 @@ class TemplateError extends Error {
...
-        const templatePath = path.join(templatesPath, pathname) + '.html';
+        const pathname = getPathName(request);
 
-        return readFile(templatePath)
-            .then((baseTemplate) => {
-                if (typeof baseTemplateFn !== 'function') {
-                    return parseTemplate(baseTemplate);
-                }
+        return getTemplatePath(templatesPath, pathname)
+            .then((templateStat) => {
```
Comment (vigneshshanmugam) [inline lib/fetch-template.js]: just a small change to keep the indentation depth minimal  
```js
.then({path} => readFile(path))
.then(baseTemplate => ....)
```

File: lib/fetch-template.js (hunk 4 above)
Comment (vigneshshanmugam) [inline lib/fetch-template.js]: Looks good to me 👍

File: lib/fetch-template.js (hunk 5 above)
Comment (luizdesign) [reply lib/fetch-template.js]: @vigneshshanmugam I understand your point of view, however i use the templateStat object to break the template parse logic on resolve the readFile promise. In your logic this not working apparently:
```js
.then(({path}) => readFile(path))
.then(baseTemplate => {
    // templateStat is not defined
})
```

We can pass the complete object to readFile, add the baseTemplate to object and resolve the promise with the object. What do you think about it?

Comment (vigneshshanmugam) [reply lib/fetch-template.js]: Ahh thanks for pointing out. No this is fine then, We can do the cleanup as part of using `util.promisify` library. 
lets get this merged :)

**Comment by vigneshshanmugam** [comment]
👍

**Comment by addityasingh** [comment]
👎

**Comment by addityasingh** [comment]
👍
